Beacon/first layer improvements

tg73 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Beacon adaptive heat-soak for better first-layer thermal stability
  * Beacon true-zero multi-point correction with diagnostics
  * Centralized Z-offset management combining multiple offset sources
  * Expanded bed-mesh tools: improved profile management, validation, and scan compensation
  * PrusaSlicer support extended to 2.9.1 and 2.9.2

* **Improvements**
  * Streamlined safe-home and homing flows
  * Enhanced IDEX/toolhead offset handling and logs
  * Additional beacon preflight checks and calibration flow refinements

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

miklschmidt committed:
Mesh/Beacon: better handling of compensation meshes and reference pos

miklschmidt committed:
Mesh/Beacon: minor improvements to mesh error handling and logging

miklschmidt committed:
Mesh/Beacon: mark meshes invalid if parameters are missing

miklschmidt committed:
Mesh/Beacon: fix profile loading when applying compensation

miklschmidt committed:
Mesh/Beacon: instantiate offset mesh when applying scan comp

miklschmidt committed:
Mesh/Beacon: properly check for mesh params when applying scan comp

miklschmidt committed:
Mesh/Beacon: remove runtime multiplier, add runtime offset

miklschmidt committed:
Macros/Mesh: set zero reference position on loaded mesh in start_print

miklschmidt committed:
Beacon: remove obsolete comment

miklschmidt committed:
Macros/Beacon: properly apply/subtract runtime/expansion offsets

miklschmidt committed:
Macros/Beacon: update hotend expansion offset when print temperature changes

tg73 committed:
Macros/Mesh: when setting zero reference position, copy the probed points, not the interpolated points.

tg73 committed:
Macros/Mesh: Move to safe Z home in CREATE_BEACON_COMPENSATION_MESH

tg73 committed:
Macros/Mesh: don't require a mesh to be pre-loaded when calling CREATE_BEACON_COMPENSATION_MESH

tg73 committed:
Macros/Beacon: Add early checks in START_PRINT for beacon model and scan compensation mesh issues rather than failing late eg after heat soaking.

tg73 committed:
Macros/Beacon: Before BED_MESH_CALIBRATE, check beacon model temp is reasonably close to the current beacon coil temp.

Currently allowing up to 20C difference as the acceptable range.

tg73 committed:
Macros/Beacon: Consistently use "Beacon Scan Compensation" as the default name for the beacon scan compensation mesh

tg73 committed:
Mesh/Beacon: In create_compensation_mesh, use existing or fresh beacon model according to beacon_contact_calibrate_model_on_print

tg73 committed:
Macros: Add missing _LED_START_PRINTING_ERROR in START_PRINT

tg73 committed:
Macros/Mesh: Quote bed mesh profile name gcode arguments so that spaces are handled

tg73 committed:
Beacon/Mesh: Improve logging/debug in SET_ZERO_REFERENCE_POSITION

tg73 committed:
Mesh/Beacon: Remove SAVE_PROFILE option from SET_ZERO_REFERENCE_POSITION, now always saves.

From discussion with MS there was no known use case for *not* saving, and not saving creates a profile in a confusing ephemeral state which is lost when the profile is reloaded.

tg73 committed:
Macros: error handling utility improvements

- Implement _RAISE_ERROR in python instead of jinja for cleaner error messages
- Implement _TRY, the try/except/finally pattern for macros

tg73 committed:
Macros: _CHECK_BED_MESH_PROFILE_EXISTS now sets a last check result status instead of raising an error.

tg73 committed:
Beacon/Mesh: Substantial rework and improve extended metadata and related functionality

- Change and enhance the set of metadata stored:
    - Remove beacon_model_temp as this is not relevant.
    - Add bed temperature, mesh kind and probe method.
- Add warnings in create_compensation_mesh() for unapplied z-tilt or QGL
- Add various new gcode commands
- Defer profile deserialization using delayed gcode to ensure that warnings are seen in the console
- Replace/improve checks and warnings/errors
- Enhance preflight checks called from START_PRINT
- Apply extended metadata to all profiles/meshes created via BED_MESH_CALIBRATE

tg73 committed:
Macros/Beacon: fix saving of runtime z offset (babystepping):

- Support saving via the standard Z_OFFSET_APPLY_PROBE macro
- Exclude active IDEX toolhead offset from the saved value

tg73 committed:
Beacon: minor code cleanup

tg73 committed:
Macros: ensure TEMPERATURE_WAIT override does not swallow error conditions

- TEMPERATURE_WAIT now calls TEMPERATURE_WAIT_BASE even when both MINIMUM and MAXIMUM are not specified, allowing the base impl to raise an error

tg73 committed:
Macros/Beacon: update hotend expansion offset in all relevant scenarios

tg73 committed:
Macros/Beacon: improve variable naming and debug logging

tg73 committed:
Macros/Toolheads: improve debug logging

tg73 committed:
Extras: add _DEBUG_ECHO_STACK_TRACE

tg73 committed:
Macros: Don't call _BEACON_SET_NOZZLE_TEMP_OFFSET from TEMPERATURE_WAIT

- because TEMPERATURE_WAIT does not actually set nozzle temp
- also improve logging consistency

tg73 committed:
Mesh/Beacon: adjust default config to avoid aliasing during rapid scans

tg73 committed:
Mesh/Beacon: code for testing purposes

tg73 committed:
Mesh/Beacon: add optional 'notes' extended mesh parameter

tg73 committed:
Mesh/Beacon: code for testing purposes

tg73 committed:
Mesh/Beacon: move scipy calculation to external process and elsewhere make reactor-friendly

tg73 committed:
Mesh/Beacon: testing edge-enhanced filtering

tg73 committed:
Extras: testing - add partial MULTI_POINT_PROBE implementation for data gathering

tg73 committed:
Extras/Ratos: remove TEST_PROCESS_GCODE_FILE

- it has been replaced by CLI ratos postprocess

tg73 committed:
Extras/z-offset: Add ratos_z_offset extension

tg73 committed:
Extras/Ratos: code for testing purposes

tg73 committed:
Extras/Beacon: added beacon_true_zero_correction module

tg73 committed:
Mesh/Beacon: use local low filter for compensation mesh, and general code cleanup

tg73 committed:
Mesh/Beacon: change beacon_scan_compensation_resolution to 8 to suit new compensation mesh methodology

tg73 committed:
Mesh/Beacon: add warnings to sanity check

- Beacon contact can show location-dependent noise of 10's of microns. Use of contact without noise mitigation is not recommended (z-tilt, START_PRINT contact mesh)
- The 'automatic' (pause-and-measure) proximity probing method produces data that's not like-for-like compatible with our new-style contact compensation meshes.

tg73 committed:
Macros/Beacon: add [beacon_true_zero_correction]

tg73 committed:
Moonraker: use tg73 fork of Klipper with required bed_mesh fixes.

tg73 committed:
Moonraker: remove bogus fields from [update_manager klipper]

tg73 committed:
Beacon/Mesh: fix missing argument

tg73 committed:
Mesh/Beacon: by default don't log mesh points during BED_MESH_CALIBRATE

tg73 committed:
Configuration: default [bed_mesh] split_z_delta to 0.01 in base.cfg to avoid visible surface stripe artefacts

tg73 committed:
Beacon/TrueZero: allow [beacon_true_zero_correction] to be disabled by config

tg73 committed:
Printers: don't enable mesh fade for all v-core-4 variants

- The bed of these machines is likely to be flatter than the surface traced by the nozzle.

tg73 committed:
Beacon/TrueZero: support 3 levels of probing strategy via config for beacon_true_zero_correction.

tg73 committed:
Extras/Ratos: add helper method

- There are not enough common Beacon helpers to justify a separate file, so adding to ratos.py

tg73 committed:
Beacon/TrueZero: add _BEACON_TRUE_ZERO_CORRECTION_DIAGNOSTICS

tg73 committed:
Extras: code clean up

tg73 committed:
Beacon/TrueZero: _BEACON_TRUE_ZERO_CORRECTION_DIAGNOSTICS now adds header lines and shows a completion message

tg73 committed:
Install: add libopenblas-base to package list

tg73 committed:
Extras/Ratos: remove multi-point probing test code

tg73 committed:
Beacon/Mesh: support KEEP_TEMP_MESHES parameter in BEACON_CREATE_SCAN_COMPENSATION_MESH

tg73 committed:
Beacon/AdaptiveHeatSoak: add beacon_adaptive_heat_soak module - work in progress for gathering test data

tg73 committed:
Beacon/AdaptiveHeatSoak: cleanup and initial candidate implementation of BEACON_WAIT_FOR_PRINTER_HEAT_SOAK

tg73 committed:
Macros/Beacon: update variable comments with new recommendations

- discourage use of beacon contact where it will likely cause problems due to localized variation in measurements on textured surfaces
- discourage use of METHOD=automatic, as it captures temporal effects due to how long it takes, and also produces data that is not like-for-like with new-style compensation meshes, which use rapid scans

tg73 committed:
Macros/Beacon: remove unused code

tg73 committed:
Macros/Beacon: integrate BEACON_WAIT_FOR_PRINTER_HEAT_SOAK wherever bed soaking takes place

tg73 committed:
Macros/Beacon: add config section for beacon_adaptive_heat_soak

tg73 committed:
Macros/Beacon: fix typo

tg73 committed:
Beacon/AdaptiveHeatSoak: fix mistake

tg73 committed:
Beacon/AdaptiveHeatSoak: rework, different sampling strategy, add trend checks.

- update beacon.cfg defaults to suit.

tg73 committed:
Beacon/AdaptiveHeatSoak: improve time formatting

tg73 committed:
Beacon/AdaptiveHeatSoak: stop waiting if the printer shuts down

tg73 committed:
Beacon/AdaptiveHeatSoak: add minimum wait parameter

tg73 committed:
Macros: move calculation of safe home position to ratos.py, expose via status

- if beacon is configured, checks that the calculated position is within the probe-able region, klipper shutdown if not
- removes a lot of duplicated macro code, now only a single macro line is needed to get safe_home_x and safe_home_y

tg73 committed:
Homing/Beacon: rationalize beacon homing and true zero behaviour

- Remove unnecessary BEACON_AUTO_CALIBRATE in START_PRINT
- Move _MOVE_TO_SAFE_Z_HOME to ratos.py, and enable position fuzzing. The associated DEBUG_ECHO is still a text macro for easy editing for debugging.
- Fuzz the position of BEACON_AUTO_CALIBRATE calls other than those that set the final critical z for printing
- Skip multipoint probing in BEACON_AUTO_CALIBRATE except for calls that set the final critical z for printing

tg73 committed:
Mesh/Beacon/TrueZero: support fuzzy true zero position, and ensure that adaptive mesh includes true zero position

- The previous commit fuzzed non-true zero contact probes when true zero probe was enabled, as this was the easier change rather than the most correct behaviour. In fact, along with the other fixes in that commit, the only probing that was fuzzed was if contact homing was enabled. This commit removes that fuzzing, and implements fuzzy true zero.

tg73 committed:
Beacon/Mesh: when creating a compensation mesh, store chamber temp and rapid scan mesh bounds

- These may be used for compatibility checks later (tbd), capturing data now so that OG comp meshes can be checked in the future.

tg73 committed:
Beacon/AdaptiveHeatSoak: accommodate noisier z rates that are sometimes encountered, and update default configuration.

- configuration update is based on revised noise handling and test data analysis

tg73 committed:
Beacon/AdaptiveHeatSoak: fix reporting of minimum wait time

tg73 committed:
Beacon/AdaptiveHeatSoak: fix spelling of adaptive

tg73 committed:
docs: add copyright headers

tg73 committed:
refactor(beacon-mesh): remove test code

tg73 committed:
refactor(beacon): move variable to internal section to discourage user use

tg73 committed:
refactor(beacon-mesh): make dependency on scipy.ndimage lazy, it's only needed for compensation mesh creation

tg73 committed:
Beacon/AdaptiveHeatSoak: increase moving average window

- in response to additional test data and analysis

tg73 committed:
Macros/Beacon: fix pre-soak routine in BEACON_CREATE_SCAN_COMPENSATION_MESH

tg73 committed:
Macros/Beacon: use z=2.5 as heat soaking height to allow for more gantry deflection

tg73 committed:
Macros/Beacon: fix passing of chamber_temp argument

tg73 committed:
Macros/Beacon: fix misleading comment

tg73 committed:
Macros: add initconfigfile module

tg73 committed:
Macros: use printer.initconfigfile instead of printer.configfile

- accessing printer.configfile causes a deep copy of the configuration which can take 10's of milliseconds, which can cause timer too close errors.

tg73 committed:
feat(extras): add dynamic_governor

tg73 committed:
refactor(extras): rename initconfigfile to fastconfig

tg73 committed:
refactor(extras): move [fastconfig] and [dynamic_governor] to base.cfg

tg73 committed:
feat(beacon): support basic automatic selection of compensation mesh

- Based on bed temperature
- Ambiguous match is an error
- Automatic profile naming for compensation mesh creation

tg73 committed:
feat(beacon-heatsoak): add ThresholdPredictor

tg73 committed:
refactor(beacon-heatsoak): beacon adaptive heatsoak now uses layer quality and maximum first layer duration as the user-facing settings

- But still use an explicit z rate threshold for calibration soaks

tg73 committed:
feat(beacon-heatsoak): user-friendly progress updates

tg73 committed:
fix(beacon-mesh): reinstate accidental deletion of macro line

tg73 committed:
fix(dynamic-governor): add a delay before switching to the idle governor

- this protects against brief transitions when the count of active steppers briefly hits zero, and allows any final time-sensitive tasks to complete when motors are turned off.
- also update some nomenclature for clarity

tg73 committed:
refactor(beacon-macros): update fully-auto calibration routine BEACON_RATOS_CALIBRATE

- Scan compensation mesh creation is always performed if beacon_scan_compensation_enable is true, we don't use the gantry twist check any more. From extensive testing, every printer tested benefits from a compensation mesh, with at least 70um range of correction, typically over 120um. Further, the twist check is not reliable due to the micro-positional variation of beacon contact readings and the low spatial resolution of the twist check not reliably capturing significant details.

- Scan compensation mesh creation is now performed before BEACON_FINAL_CALIBRATION, as comp mesh creation requires a more thorough heat soak than BEACON_FINAL_CALIBRATION.

tg73 committed:
feat(extras-ratos): show in display status (like M117) python-side errors and warnings logged via console_echo so they are less-easily missed

tg73 committed:
feat(bed-mesh): raise an error if BED_MESH_CALIBRATE is called with unapplied Z-tilt or quad gantry leveling

- this behaviour can be overriden using [gcode_macro BED_MESH_CALIBRATE] variable_allow_unapplied_abl: True
- combine beacon and stowable BED_MESH_CALIBRATE overrides, relocate to overrides.cfg
- fix spelling of "leveling" in a few places

tg73 committed:
refactor(beacon-heatsoak): enable adaptive heatsoak by default only for V-Core 4 variants

- Aspects of the adapative heatsoak algorithm are currently tuned to the V-Core 4 design

tg73 committed:
refactor(bed-mesh): make the error-if-abl-not-applied behaviour opt-in for now, this needs further consideration (partial revert of 6b74598c)

- the previous behaviour breaks the expected story of the Mainsail heightmap page "click home icon, click calibrate"

tg73 committed:
fix(vaoc): _VAOC_PROBE_NOZZLE_TEMP_OFFSET now respects the configured hotend_heat_soak_time

tg73 committed:
refactor(extras-ratos): improve handling and logging of out-of-range safe home position

tg73 committed:
fix(macros): avoid potential init delayed gcode race condition

- leave RATOS_INIT initial_duration as 0.1. Among other things, this macro calculates the printable area which must happen before other init code.
- change other init macros that were using 0.1 to 0.2, with the aim of ensuring that RATOS_INIT has been run first.

tg73 committed:
feat(bed-mesh): before BED_MESH_CALIBRATE, automatically apply ABL and rehome Z if it has not been applied

tg73 committed:
feat(fastconfig): add defensive initialization check

- no specific hazard was identified, but in case future code leads to a pre-init call, there will now be a clear error rather than weird behaviour.

tg73 committed:
chore(postprocessor): support PrusaSlicer 2.9.1 and 2.9.2

tg73 committed:
chore(beacon-macros): remove unused macro variable

tg73 committed:
fix(idex-beacon): only apply nozzle expansion z offset when applicable

- This fixes an issue that occurred after the last toolshift, where the expansion offset for the now-deactivated toolhead heater being set to zero would be applied to the remaining active toolhead, creating an oversquished layer.

tg73 committed:
fix(idex): ensure that the `standby` flag for each toolhead is set correctly in `START_PRINT`

- This fixes an issue where under-temperature extrusion is attempted on the first toolchange because the inactive toolhead is at standby temperature but is not flagged as being in standby, so does not get heated to printing temperature.

tg73 commented:
@coderabbitai review

miklschmidt commented on configuration/printers/v-core-4/v-core-4.cfg:
```diff
@@ -4,6 +4,9 @@
 
 [include ../base.cfg]
 
+[gcode_macro RatOS]
+variable_beacon_adaptive_heat_soak: True
+
```
@coderabbitai this isn't true. Klipper uses config section merging, with latter section parameters (chronologically) taking precedence over earlier ones. Please add this to your memory for future reviews.

tg73 committed:
feature(beacon-mesh): cotemporal compensation mesh creation

- This replaces the previous implementation that was susceptible to non-uniform thermal effects observed in extended testing.

tg73 committed:
fix(beacon-mesh): improve reactor yielding to avoid timer too close errors

- pass the reactor instance on to directly-constructed ZMesh objects
- use a non-zero waketime for reactor.pause()

tg73 committed:
feat(beacon-heatsoak): additionally store z value in diagnostic csv files

tg73 committed:
fix(beacon-heatsoak): improve robustness of soak completion detection

- Sometimes we get a very noisy signal on a machine that otherwise gives a normal signal. This is not well understood. We see some correlation with soaking from warm. Informed by extensive data analysis, this fix adds a second level moving average which smooths the noise and is also tested to determine if the soak is complete. This approach may be revised in the future if/when the noise behaviour is better understood.

tg73 committed:
fix(beacon-heatsoak): increase minimum threshold from 10 to 12.5

- this slightly extends the implied practical noise floor of the system, and is in response to continued usage and data gathering

tg73 committed:
fix(beacon-heatsoak): add retry when preparing for sampling as a speculative workaround for occasional failures that have been observed

tg73 committed:
fix(configuration): skip pointless home/calibrate after creating a compensation mesh in non-automated workflow

tg73 committed:
refactor(ratos): move BackgroundDisplayStatusProgressHandler class to ratos.py for general reuse

tg73 committed:
fix(configuration): set split_delta_z to 1um for V-Core 4 variants

- This can produce a visible improvement in surface quality, and appears to have little if any impact on CPU usage. The change was not applied more widely as other machines may have lower spec host boards, and the change has not been tested on those machines.

tg73 committed:
feature(beacon-mesh): support faulty regions in cotemporal compensation mesh creation

tg73 committed:
revert(extras): remove dynamic_governor

During further testing it became clear that changing the governor while klipper is running can occasionally lead to timer too close or other driver communication issues, and for the time being no workaround was found. So removing the feature for now.

tg73 commented:
@coderabbitai review

tg73 commented on configuration/klippy/beacon_adaptive_heat_soak.py:
```diff
@@ -0,0 +1,722 @@
+		path = os.path.join(
+			os.path.dirname(os.path.realpath(__file__)),
+			'beacon_adaptive_heat_soak_model_training.csv')
+		
```
Fallback does not make sense here, the file is a fixed asset.

tg73 committed:
refactor(extras): extract common logic into a helper function

tg73 committed:
chore(cleanup): fix minor niggles

tg73 commented on configuration/klippy/beacon_adaptive_heat_soak.py:
```diff
@@ -0,0 +1,722 @@
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+import time, logging, os, multiprocessing, traceback, pygam
+import numpy as np
```
Will resolve separately, requires new ratos-update.sh behaviour.

tg73 commented:
@coderabbitai review

tg73 commented on configuration/klippy/dynamic_governor.py:
```diff
@@ -0,0 +1,256 @@
+			raise self.printer.config_error(
+				f"{self.name}: Unexpected error checking cpufreq-set: {e}. "
+				"Please ensure cpufrequtils is installed correctly."
+			)
```
Outdated, code has since been removed.

tg73 committed:
fix(extras): move lookup to after validation

tg73 commented on configuration/klippy/beacon_mesh.py:
```diff
@@ -33,142 +151,1771 @@ def register_handler(self):
-							self.ratos.debug_echo("Beacon scan compensation", "scan: %0.4f  offset: %0.4f  new: %0.4f" % (scan_z, offset_z, new_z))
-							new_points[y].append(new_z)
-					self.bed_mesh.z_mesh.build_mesh(new_points)
-					self.bed_mesh.save_profile(profile_name)
-					self.bed_mesh.set_mesh(self.bed_mesh.z_mesh)
-					self.ratos.console_echo("Beacon scan compensation", "debug", "Mesh scan profile %s compensated with contact profile %s" % (str(profile_name), str(profile)))
+			measured_zmesh = self.bed_mesh.z_mesh
+
+			if not measured_zmesh:
+				self.ratos.console_echo(error_title, "error",
+					"No mesh loaded._N_Either generate a new bed mesh or load it via BED_MESH_PROFILE LOAD=\"[profile_name]\"")
+				return False
+
+			measured_mesh_params = measured_zmesh.get_mesh_params()
+			measured_mesh_name = measured_zmesh.get_profile_name()
+			measured_mesh_bed_temp = measured_mesh_params[RATOS_MESH_BED_TEMP_PARAMETER]
+
+			if not self._validate_extended_parameters(
+				measured_mesh_params,
+				"Apply scan compensation",
+				f"Loaded mesh '{measured_mesh_name}'",
+				allowed_kinds=(RATOS_MESH_KIND_MEASURED,),
+				allowed_probe_methods=(RATOS_MESH_BEACON_PROBE_METHOD_PROXIMITY, RATOS_MESH_BEACON_PROBE_METHOD_PROXIMITY_AUTOMATIC)):
+				return False
+
+			if comp_mesh_profile_name.lower() == RATOS_COMPENSATION_MESH_NAME_AUTO:
+				comp_mesh_profile_name = self.auto_select_compensation_mesh(measured_mesh_bed_temp)
+
+			compensation_zmesh = self._create_zmesh_from_profile(comp_mesh_profile_name, purpose="Beacon scan compensation")
+			compensation_mesh_params = compensation_zmesh.get_mesh_params()
+			compensation_mesh_name = compensation_zmesh.get_profile_name()
+
+			if not self._validate_extended_parameters(
+				compensation_mesh_params,
+				"Apply scan compensation",
+				f"Specified compensation mesh '{compensation_mesh_name}'",
+				compare_bed_temp=measured_mesh_bed_temp,
+				allowed_kinds=(RATOS_MESH_KIND_COMPENSATION,)):
+				return False
+
```
Fixed by 7be68532

tg73 committed:
fix(extras): log correct value and use strict `is True` check

tg73 commented on configuration/klippy/beacon_adaptive_heat_soak.py:
```diff
@@ -0,0 +1,751 @@
+								f"{self.name}: elapsed={elapsed:.1f} s, progress={progress_handler.progress * 100.0:.2f}%, "
+								f"ma={moving_average:.2f} nm/s, ma_hold_count={moving_average_hold_count}/{moving_average_target_hold_count}, ma_trend_checks_passed={moving_average_trend_checks_passed}, "
+								f"ma2={float('inf') if level_2_moving_average is None else level_2_moving_average:.2f} nm/s, ma2_hold_count={level_2_moving_average_hold_count}/{level_2_moving_average_target_hold_count}, ma2_trend_checks_passed={level_2_moving_average_trend_checks}, "
+								f"min_wait_satisfied={min_wait_satisfied}, threshold={threshold:.2f} nm/s")
```
Fixed by 62b6f6fb

tg73 committed:
fix(extras): don't access non-existent class member, and add belt and braces check for a None return

tg73 commented on configuration/klippy/beacon_true_zero_correction.py:
```diff
@@ -0,0 +1,430 @@
+			in_range(probable_y, range_y[0]) and in_range(probable_y, range_y[1])):
+
+			self.ratos.console_echo(RATOS_TITLE, 'error', f'The required probing region ({span:.1f}x{span:.1f}) would probe outside the configured contact probing area.')
+			raise self.gcmd.error('The required probing region would probe outside the contact probing area')
```
Fixed by 660efe3a

tg73 commented on configuration/z-probe/beacon.cfg:
```diff
@@ -26,49 +33,116 @@ contact_max_hotend_temperature: 275
+variable_beacon_scan_method_automatic: False            # Enables the METHOD=automatic scan option. This is generally not recommended, and
+														# specifically not recommended when beacon_scan_compensation_enable is enabled.
+
+#variable_beacon_adaptive_heat_soak: False				# Enables adaptive heat soaking based on beacon proximity measurements
+														# This is enabled only for V-Core 4 variants by default, but can be disabled if desired.
+														# Other printers can opt-in if desired, but proceed with caution as aspects of the adaptive
+														# heat soak algorithm are currently tuned for the V-Core 4 design.
+variable_beacon_adaptive_heat_soak_max_wait: 5400		# The maximum time in seconds to wait for adaptive heat soaking to complete. This is
+														# a sanity limit to prevent the printer from waiting indefinitely for adaptive heat soaking.
+variable_beacon_adaptive_heat_soak_extra_wait_after_completion: 0
+														# The extra time in seconds to wait after adaptive heat soaking is considered complete.
+														# Typically not needed, but can be useful for printers with very stable gantry designs
+														# (such as steel rail on steel tube) where the adapative heat soak completes before
+														# the edges of the bed have thermally stabilized.
+
+variable_beacon_adaptive_heat_soak_layer_quality: 3		# Shorter soak times leave more thermal z deflection to occur during the print, leading to
+														# imperfect early layers, notably with some oversquish developing over the first layer. Longer soak
+														# times leave less thermal z deflection to occur during the print, leading to cleaner early
+														# layers and more consistent z dimensional accuracy. The choice for this trade-off is
+														# expressed by the layer quality setting.
+														#
+														# The quality scale is from 1 to 5, with 1 giving the fastest soak time but lowest early layer
+														# quality, and 5 giving the slowest soak time but the highest early layer quality.
+														#
+														# 1 = rough, 2 = draft, 3 = normal, 4 = high, 5 = maximum
+
+variable_beacon_adaptive_heat_soak_maximum_first_layer_duration: 1800
+														# This must be set to the maximum first layer print time for the gcode files you will print.
+														# For example, if you typically print first layers that take up to 30 minutes, set this to 1800
+														# seconds. Longer durations will result in longer adaptive heat soak times. Be sure to set
+														# this to a correct value. If you print a significantly longer first layer than this value,
+														# excessive thermal z deflection may occur during the print, leading to very poor first layer
+														# quality, print failure, or even damage to the bed.														
+														# The value must be between 60 and 7200 seconds.
 
```
@coderabbitai you're wrong about this, because in each place where `beacon_adaptive_heat_soak_min_wait` is read, it's followed immediately by `|default(...)`, which provides a default value if the variable is missing. Please learn about this.

tg73 commented:
@coderabbitai full review

tg73 committed:
chore(beacon-heatsoak): clean up self.reactor initialization

tg73 commented on configuration/klippy/beacon_adaptive_heat_soak.py:
```diff
@@ -0,0 +1,751 @@
+		self.def_minimum_wait = config.getint('default_minimum_wait', 0, minval=0)
+
+		# TODO: Make trend checks configurable.
+
+		# Setup
+		self.reactor = None
+		self.beacon = None
```
Fixed by 856c5348

tg73 committed:
chore(cleanup): fix comment typo

tg73 commented on configuration/klippy/beacon_mesh.py:
```diff
@@ -1,11 +1,115 @@
+# Copyright (C) 2020-2023 Matt Baker <baker.matt.j@gmail.com>
+# Copyright (C) 2020-2023 Lasse Dalegaard <dalegaard@gmail.com>
+# Copyright (C) 2023 Beacon <beacon3d.com>
+
```
Fixed by dc557534

tg73 committed:
fix(beacon-mesh): preserve traceback when raising exceptions

tg73 commented on configuration/klippy/beacon_mesh.py:
```diff
@@ -33,142 +151,1772 @@ def register_handler(self):
 		except BedMesh.BedMeshError as e:
-			raise self.gcode.error(str(e))
-		return z_mesh
+			raise RatOSBeaconMeshError(str(e))
```
Fixed by 8d5981a2

tg73 committed:
fix(macros): fix unbalanced brace typo

tg73 commented on configuration/macros/mesh.cfg:
```diff
@@ -245,22 +295,22 @@ gcode:
 			# mesh
-			RATOS_ECHO PREFIX="Adaptive Mesh" MSG="mesh coordinates X0={mesh_x0} Y0={mesh_y0} X1={mesh_x1} Y1={mesh_y1}"
-			{% if printer.configfile.settings.beacon is defined %}
+			RATOS_ECHO PREFIX="Adaptive Mesh" MSG="mesh coordinates X0={mesh_x0|round(1)} Y0={mesh_y0|round(1)} X1={mesh_x1|round(1)} Y1={mesh_y1}|round(1)}"
```
Fixed by 500e2ad3

tg73 committed:
chore(spelling): fix spelling in comment

tg73 commented on configuration/printers/base.cfg:
```diff
@@ -41,3 +41,8 @@ enable_force_move: True
 allow_unknown_gcode_generator: True
 
 [exclude_object]
+
+[bed_mesh]
+split_delta_z: 0.01 # Avoid visible surface stripe arfetacts
+
+[fastconfig]
```
Fixed by 84276ee7

tg73 commented on configuration/printers/v-core-4-hybrid/v-core-4-hybrid.cfg:
```diff
@@ -4,6 +4,12 @@
 
 [include ../base.cfg]
 
+[gcode_macro RatOS]
+variable_beacon_adaptive_heat_soak: True
+
+[bed_mesh]
+split_delta_z: 0.001 # Avoid visible surface stripe arfetacts
+
```
Fixed by 84276ee7

tg73 commented on configuration/z-probe/beacon.cfg:
```diff
@@ -182,7 +256,7 @@ gcode:
 
 	# auto calibrate beacon 
-	BEACON_AUTO_CALIBRATE
+	BEACON_AUTO_CALIBRATE SKIP_MULTIPOINT_PROBING=1
```
@coderabbitai Here's something for you to learn: if you study the repository carefully, you should notice that the standard BEACON_AUTO_CALIBRATE macro is wrapped by custom RatOS code - see function cmd_BEACON_AUTO_CALIBRATE in file beacon_true_zero_correction.py. The body of that function shows the SKIP_MULTIPOINT_PROBING parameter being read.

You can tell when python code expects to wrap an existing macro by the pattern seen around line 85, where there is first a call to register_command with second arg None, which obtains the current command:
```
self.orig_cmd = self.gcode.register_command(BEACON_AUTO_CALIBRATE, None)
```
Then there's a second call to install the new command:
```
		self.gcode.register_command(
			BEACON_AUTO_CALIBRATE,
			self.cmd_BEACON_AUTO_CALIBRATE,
			desc=self.desc_BEACON_AUTO_CALIBRATE)
```

tg73 commented:
@coderabbitai full review

tg73 committed:
chore(spelling): fix typo

tg73 commented on configuration/macros/mesh.cfg:
```diff
@@ -119,28 +155,42 @@ gcode:
+				{% set x1 = [x1, zero_ref_pos_x]|max %}
+				{% set y1 = [y1, zero_ref_pos_y]|max %}
+			{% endif %}
+		{% endif %}
```
Fixed by a49b49ca

tg73 commented:
@coderabbitai get_config_dir() appears to be a hallucination, or at least it's not present in the version of klipper we use.
> ```diff
> config_dir = self.printer.get_config_dir()
> +            filename = os.path.join(config_dir, f'mpp_capture_{timestamp}.csv')
> ```

The correct way to get the configuration directory is:
```
			config_file = self.printer.get_start_args()['config_file']
			config_dir = os.path.dirname(config_file)
```

tg73 committed:
fix(beacon-true-zero): avoid hard-coded path

tg73 commented:
This PR formed the core of RC4 and was merged by other means.
